src/codec.rs

```diff
@@ -1113,71 +1067,110 @@ impl<T: Encode> Encode for [T] {
 	}
 }
 
-/// Create a `Vec<T>` by casting directly from a buffer of read `u8`s
-///
-/// The encoding of `T` must be equal to its binary representation, and size of `T` must be less or
-/// equal to [`MAX_PREALLOCATION`].
-pub(crate) fn read_vec_from_u8s<I, T>(input: &mut I, items_len: usize) -> Result<Vec<T>, Error>
+fn decode_vec_chunked<T, F>(len: usize, mut decode_chunk: F) -> Result<Vec<T>, Error>
 where
-	I: Input,
-	T: ToMutByteSlice + Default + Clone,
+	F: FnMut(&mut Vec<T>, usize) -> Result<(), Error>,
 {
 	debug_assert!(MAX_PREALLOCATION >= mem::size_of::<T>(), "Invalid precondition");
```

Comment:
We should make this into a static assert and check it at compile time.

Reply:
I couldn't manage to do this so far. I tried something like

```
const _: () = {
    assert!(MAX_PREALLOCATION >= mem::size_of::<T>())
}
```

inside `decode_vec_chunked()`

But I'm getting an error: `can't use generic parameters from outer item`.

Any suggestion would be helpful

Reply:
You don't need to define a constant; since Rust 1.79 you can use a `const {}` block to force const evaluation of an expression.

Reply:
Yes, this works, thanks ! PTAL on #615 

But the CI fails, because the CI image uses rust 1.73.0 .  We can either try to release a `paritytech/ci-unified:bullseye-1.79.0` image. Will check how this can be done.